### DISCUSSION TITLE
how to input massive amount data to the map-reduce example

### DISCUSSION THREAD

**Discussion by aixi** [argoproj/argo-workflows#12204] [Q&A] [answered]
I found the https://github.com/argoproj/argo-workflows/blob/master/examples/map-reduce.yaml example, which perfectly match my workflow pattern, but it can only process one json file, is there a way to  input more data, can I use a s3 folder or a kafka topic as split node inputs. Is there a working yaml example.
thank you.

**Comment by aixi** [comment]
I don't think it's possible, argo is not flink

**Reply by agilgur5** [reply]
Argo is not _specifically_ a stream processing framework like Flink and so is not necessarily optimized for certain use-cases.
It is a _generic_ Workflow engine which can be used for a variety of use-cases. [That example](https://github.com/argoproj/argo-workflows/blob/master/examples/map-reduce.yaml) shows how you can apply it to a map-reduce problem.

There is a sibling project, from some Intuit folks that worked on Events and Workflows, [Numaflow](https://github.com/numaproj/numaflow), that _is_ optimized specifically for the stream processing use-case, if that's what you're looking for.

**Comment by agilgur5** [comment]
> but it can only process one json file, is there a way to input more data, can I use a s3 folder

[That example](https://github.com/argoproj/argo-workflows/blob/master/examples/map-reduce.yaml) _is_ using an S3 folder: `{{workflow.name}}/parts/{{item}}.json`. `parts` is the folder and each `item` is a part

**Reply by aixi** [reply]
Thank you for your reply.
I already using this example, it works perfectly for me as a offline data processing system.
What bothers me is I need a online service require a little bit of realtime, I already have a complex DAG, it can running on argo workflow, it input data from object storage system, but now i want to make my workflow running as a long running server, which input data from http or grpc request. I make a try, but it's not working, may its not the right use pattern, I will look into the numaflow

**Reply by agilgur5** [reply, marked as answer]
So [Argo Events](https://github.com/argoproj/argo-events) was created to handle event-driven Workflows -- it tightly integrates with Workflows and the Workflows UI still has certain pages specifically for Events (I think Events does not have its own independent UI; there's some history around that since Workflows was the first Argo project).
The Events Lead is also a Lead for Numaflow, as far as I know.

So if you already have Workflows as part of your infra, adding Events may make sense. If you have a more specific use-case like streaming and aren't planning to add more use-cases for Workflows, then Numaflow may serve your purposes better. I've never used Numaflow though, so I can't really comment on it more than that.
I believe it has better performance with regard to real-time, as Argo Workflows's current architecture (closely tracked, complex DAGs, plus new Pods for each task that can run entirely different images) can make low latency a bit challenging (not impossible, but not necessarily easy). Numaflow is more homogenous in its use-case, so it can be better optimized for streaming as such.

If you don't necessarily need very low latency and just need online processing, Workflows + Events can handle most things in that space

**Reply by aixi** [reply]
I have done some experiments using argo event and argo workflow already. it works good for batch processing, like user submit large amount data.But in other situation, like user submit only one data, argo workflow will create new pod to do the job, it takes minutes, that's a little slow for user.
I will look into  numaflow, maybe i could find another way
